Set FW compile date in AUTOPILOT_VERSION.os_sw_version

peterbarker opened the pull request:
We're well-OOB standards-wise on this.

But we haven't used this field for anything in the past, and it has proved extremely useful to have this information available in AUTOPILOT_VERSION.

peterbarker commented on ArduCopter/version.cpp:
```diff
@@ -28,14 +28,21 @@ const AP_FWVersion Copter::fwver{
     .fw_type = FW_TYPE,
 #ifndef GIT_VERSION
     .fw_string = THISFIRMWARE,
+    .fw_hash_str = "",
```
You get a compiler error about non-trivial designators otherwise.

OXINARF replied:
I'm a bit surprised by that it doesn't work given that GCC is supposed to have an extension to support designated initializer lists even in C++. I see there is a long standing bug open in GCC for this.

Wouldn't it be fixed though with just moving the `os_sw_version` to be declared before the uninitialized fields in the struct? Or adding a constructor to the struct.

peterbarker commented:
@OXINARF Does this really stink of "should be a mavlink 2 extension" to you?

peterbarker commented:
@OXINARF The utility?  While we understand what a git hash is, others don't.

Identifying a software release by the date it was released on seems to be easier for interactions with other people.  People can generally remember when releases went out, and they talk in terms of that release date, so supplying that number directly to them in this manner seems to be an extremely useful option.

OXINARF commented:
> Does this really stink of "should be a mavlink 2 extension" to you?

Well, I would say that looking at the message we would definitely be doing something that we shouldn't. The *X_sw_version* should be a match to the information in *X_custom_version* (I think we don't fill the *os/middleware_sw_version* just because it was too much work and not worth it to extract that information into our code).
So to answer your question, yes, I think a MAVLink2 extension would be appropriate.

>The utility? While we understand what a git hash is, others don't.
>
>Identifying a software release by the date it was released on seems to be easier for interactions with other people. People can generally remember when releases went out, and they talk in terms of that release date, so supplying that number directly to them in this manner seems to be an extremely useful option.

I understand that people relate more to a date, but if someone gives me a date as their software version I have no idea what code they are running, so I still don't see how it is useful.
We have several cases:

- our stable software releases: each vehicle has a specific version number for each release and people relate to that even more than a date
- our beta software releases: same as above
- our master builds: this is the only place I could see some utility, but if there are multiple builds in the same day then I have to go to our firmware server to check at what time they were compiled so I can find out what code was running; seems way more trouble than a user (running development code) supplying the Git hash
- user built firmware: what firmware is it? I can build Copter 3.3 today, I give you today's date, how do you know I'm running Copter 3.3?

My main objection is that I don't see how this can be used for anything. On top of that, it uses the build date, not even the Git HEAD commit date - although I still find that completely useless, it's better than the build date.

peterbarker commented:
On Fri, 1 Dec 2017, Francisco Ferreira wrote:

> Wouldn't it be fixed though with just moving the os_sw_version to be
> declared before the uninitialized fields in the struct? Or adding a
> constructor to the struct.

I did what you see first.  Then I changed it to what you suggested.  Then 
I changed it back.  It's something we need to cope with and keeping the 
field initialisation order in place seemed to be nicer.

amilcarlucas commented:
nice one 👍

peterbarker commented:
On Wed, 14 Feb 2018, Amilcar Lucas wrote:

> nice one 👍

I think our current theory might be to replace this with mavlink2 fields 
instead....

amilcarlucas commented:
needs a merge

magicrub commented:
needs a *rebase*

amilcarlucas commented:
:) yes, that as well :)

WickedShell commented:
This would make it impossible to build the same binary again unless you could/did override the value. (We actually can do that at the moment, it's a bit sad to see the idea of that going away)

auturgy commented:
Repeatable binaries are important for verification. It’s not something we should give up in my view.

Regards,

James 

> On 27 Feb 2019, at 10:31 am, WickedShell <notifications@github.com> wrote:
> 
> This would make it impossible to build the same binary again unless you could/did override the value. (We actually can do that at the moment, it's a bit sad to see the idea of that going away)
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

OXINARF commented:
I'll stand by my initial review: https://github.com/ArduPilot/ardupilot/pull/7287#pullrequestreview-79199848

peterbarker commented:
@WickedShell Sooo... override the date/time if you want to repeat the build?

Nothing @OXINARF says above is incorrect - but people can and do use dates to relate things to one-another.

So we close this issue as "the users are wrong"?

peterbarker commented:
Talked about this on the devcall.

Consensus seems to be that there is real use in something like this, and we should add it as a compile option to waf.

OXINARF commented:
> Consensus seems to be that there is real use in something like this

@peterbarker It would be nice to list what uses were identified for posterity.

peterbarker commented:
On Tue, 5 Mar 2019, Francisco Ferreira wrote:

>       Consensus seems to be that there is real use in something like this
> @peterbarker It would be nice to list what uses were identified for posterity.

Pretty much the same arguments that have been made.  People relate to 
dates, not git hashes.  Yes, it's wrong, but we live in a messy world.

It was decided this wouldn't be something that would be on by default - it 
would be a waf configure argument to enable it (and we'll have some way of 
specifying the date/time to use so you should be able to get a repeatable 
build even if using date/time).

peterbarker commented on libraries/GCS_MAVLink/GCS_Dummy.h:
```diff
@@ -6,7 +6,11 @@ const AP_FWVersion fwver
+    fw_string: "Dummy GCS",
+    fw_hash_str: "",
+    middleware_hash_str: "",
+    os_hash_str: "",
```
To avoid non-trivial initialiser errors from the compiler.
